### PR TITLE
Real life financial use case demo (Completed)

### PR DESCRIPTION
**Pull request by li-vendia**
real life financial use case demo migration

### CODE REVIEW

**Comment by li-vendia** [comment]
For reviewers, don't both with most of the files as they were ported. Focus on `demos/loan-platform/setup.sh` and any files under `terraform` folders.

**Review by james-vendia — COMMENTED** [review]
@li-vendia - Sharing some early feedback to help simplify the eventual PR.

* You can exclude the contents of the `data/loans` and `data/payments` directories from version control.  There are scripts to generate both so no need to maintain these.

* I'd move `terraform` under `src` (since it is) and then re-organize that area a bit to group similar things - just a thought `terraform`, `smart-contracts`, `webapp`, `data-setup` or something like that (you decide).

**Comment by li-vendia** [comment]
> @li-vendia - Sharing some early feedback to help simplify the eventual PR.
> 
> * You can exclude the contents of the `data/loans` and `data/payments` directories from version control.  There are scripts to generate both so no need to maintain these.
> * I'd move `terraform` under `src` (since it is) and then re-organize that area a bit to group similar things - just a thought `terraform`, `smart-contracts`, `webapp`, `data-setup` or something like that (you decide).

Could you direct me to the script that generates the data? @james-vendia 
I know the command `npm run addLoans` will load all the loans under that folder, but I didn't see any steps generating the files.

So the reason I didn't put terraform under src it's because terraform files are not part of the node.js application and is its own entity. I wanted to avoid reader confusion.

**Comment by james-vendia** [comment]
>Could you direct me to the script that generates the data? @james-vendia

Sure thing.

* https://github.com/vendia/solutions/blob/main/customers/fanniemae/loan-platform/src/PaymentGenerator.js
* https://github.com/vendia/solutions/blob/main/customers/fanniemae/loan-platform/src/LoanGenerator.js

>I know the command npm run addLoans will load all the loans under that folder, but I didn't see any steps generating the files.

* https://github.com/vendia/solutions/blob/a5398b37176e5b274927f59ffcad7342add8873e/customers/fanniemae/loan-platform/src/package.json#L13
* https://github.com/vendia/solutions/blob/a5398b37176e5b274927f59ffcad7342add8873e/customers/fanniemae/loan-platform/src/package.json#L14

> So the reason I didn't put terraform under src it's because terraform files are not part of the node.js application and is its own entity

I agree.  I would keep ALL of it under `src` and then have subdirectories to differentiate within that directory.  We have Node scripts + Terraform + React App.  I think each should be self-contained (as you mentioned) but it seems like all should be under `src` since it's all source code.

**Comment by li-vendia** [comment]
The demo is completed. For reviewers, don't bother with most of the files as they were ported. Focus on demos/loan-platform/setup.sh and any files under terraform folders. Also the README.md. Thanks

File: terraform/lambda-role.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (james-vendia) [inline terraform/lambda-role.json:1]: Are the files under `terraform` intentional or should those be gone now that `src/terraform` exists?

File: demos/loan-platform/README.md [hunk 2]
```diff
@@ -0,0 +1,173 @@
+# Loan Platform Use case
+
+# Purpose
+
+This demo is meant to mimic a real world loan platform scenario where multiple financial institutions have to share their data across organizations and make changes to loans. All names are to keep things as realistic as possible.
+
+# Pre-Requisite
```
Comment (james-vendia) [inline demos/loan-platform/README.md:7]: What AWS keypair or identity must be obtained prior to the Terraform portion of the setup script?  I may have missed the location where those are called out but I didn't see that in the write.  Am I missing something that grants the caller of the script access to "their" AWS dev environment?

**Review by james-vendia — COMMENTED** [review]
A few questions to make sure I'm following the approach and interaction between `setup.sh` and an SA's AWS dev account.

File: demos/loan-platform/README.md (hunk 2 above)
Comment (li-vendia) [reply demos/loan-platform/README.md:7]: Great point out.  A better written terraform script would have AWS_SECRETS and so in a file. But for our environment we do swap between accounts. So the way I failed to put it in is that before running `terraform apply`, we need to have our AWS CLI setup and working properly.

File: terraform/lambda-role.json (hunk 1 above)
Comment (li-vendia) [reply terraform/lambda-role.json:1]: I intentionally have an additional one at the root as a template to create for other demos.

File: demos/loan-platform/README.md (hunk 2 above)
Comment (li-vendia) [reply demos/loan-platform/README.md:7]: Updated below in README.md to address this:


Have your aws credentials in your environment variable. This is needed for the terraform script.
```
export AWS_ACCESS_KEY_ID=<your-aws-access-token>
export AWS_SECRET_ACCESS_KEY=<your-aws-accee-token-secret>
```

**Comment by david-vendia** [comment]
Hey @li-vendia , I tried to run the setup.sh, but received an error:
``
❯ terraform --version
Terraform v1.3.7
on darwin_arm64
❯ bash setup.sh
Getting your username..
Enter your email: David.McElligott+demo@vendia.net
Password: *******************
Auth Code: ******
Your username is:  for
david.mcelligott+demo@vendia.net Your nodes will be created under this username.
Copy sample file and replacing all user ids to above user id.
sed: 1: "s/you@vendia.net/for
da ...": unescaped newline inside substitute pattern
[2023-02-03T16:45:58-0600]: failed to generate registration files.
```

**Comment by david-vendia** [comment]
@li-vendia, after running setup.sh again it seems to be working.

I can try to validate everything functions when I get back!

**Comment by li-vendia** [comment]
Thanks David! Enjoy your vacation!On Feb 3, 2023, at 5:59 PM, David McElligott ***@***.***> wrote:﻿
@li-vendia, after running setup.sh again it seems to be working.
I can try to validate everything functions when I get back!

File: demos/loan-platform/setup.sh [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+export const nodes = Object.keys(config);"
+
+if [[ $? -ne 0 ]] 
+then
+  err "failed to create __config.tsx file."
+  exit 1
+fi
+
+## install vendia module and dependencies
+(cd src/app; share client:pull;)
```
Comment (david-vendia) [inline demos/loan-platform/setup.sh:240]: Could you update this command to provide the uni and node to generate the SDK for?

**Comment by li-vendia** [comment]
Yes. Let me do some reading on how to automatically input choices in a shell script. Is it from stdin?On Feb 3, 2023, at 6:34 PM, David McElligott ***@***.***> wrote:﻿
@david-vendia commented on this pull request.



In demos/loan-platform/setup.sh:
> +      '${websocket_url[5]}',
+    apiKey: '${node_keys[5]}',
+    debug: true,
+  },
+};
+
+export const nodes = Object.keys(config);"
+
+if [[ $? -ne 0 ]] 
+then
+  err "failed to create __config.tsx file."
+  exit 1
+fi
+
+## install vendia module and dependencies
+(cd src/app; share client:pull;)

Could you update this command to provide the uni and node to generate the SDK for?

File: demos/loan-platform/setup.sh (hunk 3 above)
Comment (li-vendia) [reply demos/loan-platform/setup.sh]: Done. The new command is 
```
(cd src/app; printf "y" | share client:pull --uni $uni_name --node $nodes_names[4];)
```

**Review by AVukovic-Vendia — APPROVED** [review]

**Pull request merged**